pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -462,6 +467,74 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileBuildRun) updateBuildRunUsingTaskRunCondition(ctx context.Context, buildRun *buildv1alpha1.BuildRun, taskRun *v1beta1.TaskRun, trCondition *apis.Condition) error {
+	var reason, message string = trCondition.Reason, trCondition.Message
+
+	switch v1beta1.TaskRunReason(reason) {
+	case v1beta1.TaskRunReasonTimedOut:
+		reason = "BuildRunTimeout"
+		message = fmt.Sprintf("BuildRun %s failed to finish within %s",
+			taskRun.GetLabels()[buildv1alpha1.LabelBuildRun],
+			taskRun.Spec.Timeout.Duration,
+		)
+
+	case v1beta1.TaskRunReasonFailed:
+		if taskRun.Status.CompletionTime != nil {
+			var pod corev1.Pod
+			if err := r.client.Get(ctx, types.NamespacedName{Namespace: taskRun.Namespace, Name: taskRun.Status.PodName}, &pod); err != nil {
+				return err
+			}
+
+			// Since the container status list is not sorted, as a quick workaround mark all failed containers
+			var failures = make(map[string]struct{})
+			for _, containerStatus := range pod.Status.ContainerStatuses {
+				if containerStatus.State.Terminated != nil && containerStatus.State.Terminated.ExitCode != 0 {
+					failures[containerStatus.Name] = struct{}{}
+				}
+			}
+
+			// Find the first container that failed
+			var failedContainer *corev1.Container
+			for i, container := range pod.Spec.Containers {
+				if _, has := failures[container.Name]; has {
+					failedContainer = &pod.Spec.Containers[i]
+					break
+				}
+			}
```

Comment:
The TaskRun status contains the containers as well and there they should be sorted.

Reply:
Are u referring to the `taskRun.Status.Steps` ? If this is the case, the steps do not contain a "name" that we can use to produce a reference to the failing step. Not sure why they are empty, they only have the `corev1.ContainerState` populated.

Reply:
Yes, I am referring to `taskRun.Status.Steps` and I do see the `name` which is the step name from the tekton task as well as the `container` which includes the `step-` prefix that Tekton prepends to create a container name. Example:

```yaml
    - container: step-build-and-push
      imageID: icr.io/obs/codeengine/kaniko/executor@sha256:b9eec410fa32cd77cdb7685c70f86a96debb8b087e77e63d7fe37eaadb178709
      name: build-and-push
      terminated:
        containerID: containerd://a343919803372cdb214bb0456c02ec01479f2d8fc99188ef1ec42220783db858
        exitCode: 1
        finishedAt: "2020-12-01T19:35:14Z"
        reason: Error
        startedAt: "2020-12-01T19:34:47Z"
```

But if you have an example where this is not the case, we will have to look at the pod.

Reply:
And if we stick to loading the pod, do we need to handle the case that the pod has been deleted?

Reply:
For `taskRun.Status.Steps`, this is not reliable, because also Tekton populates them from the `pod.Status.ContainerStatuses`. I think the pod is the source of truth. If we rely on `taskRun.Status.Steps` we might query this while the object have not been fully populated, therefore it is prompt to errors.

For `do we need to handle the case that the pod has been deleted?` . I think this is only concerning if the previous state of the of the BuildRun condition stills reflects a "Running", while the TaskRun is already "Failed". Therefore is absolutely important to handle this, but in my opinion is a very edge case. Is a very edge case, because in order for this to happen, it means that the pod deletion took place between `a TaskRun status was updated, e.g. from Running to Failed` and `during the BuildRun reconciler reacts on that TaskRun state change and reach the code when the pod is retrieved`(we are speaking on miliseconds here)

Proposal to handle the above will look like:

```go
	case v1beta1.TaskRunReasonFailed:
		if taskRun.Status.CompletionTime != nil {
			var pod corev1.Pod
			if err := r.client.Get(ctx, types.NamespacedName{Namespace: taskRun.Namespace, Name: taskRun.Status.PodName}, &pod); err != nil {
				if apierrors.IsNotFound(err) {
					message = fmt.Sprintf("buildrun failed, pod %s not found",
						taskRun.Status.PodName)
					break
				}
				return err
			}
```

opinions?

Reply:
Changes look good. I managed our build controller to reconcile a failed taskrun with a pod already deleted:

```yaml
    conditions:
    - lastTransitionTime: "2020-12-17T12:47:00Z"
      message: buildrun failed, pod sascha-npm-simple-buildpacks-1-wpzpt-pod-lckx9 not found
      reason: Failed
      status: "False"
      type: Succeeded
```

Reply:
I did this also, but I was not able to put this on a test case, only manual hack, how you did this?

Reply:
The only way I was able to get this is this test:

1. Create a build run
2. Wait for it to be running
3. Terminate the build operator
4. Delete the pod
5. Wait for the TaskRun to be failed
6. Start the build operator again
-> it now starts up and reconciles the running build run, updates it with the failed taskrun and cannot find a pod

Reply:
Thanks :)